debezium-core/src/main/java/io/debezium/converters/spi/RecordParser.java

```diff
@@ -39,20 +41,31 @@ public abstract class RecordParser {
             AbstractSourceInfo.SNAPSHOT_KEY,
             AbstractSourceInfo.DATABASE_NAME_KEY);
 
-    protected RecordParser(Schema schema, Struct record, String... dataFields) {
-        this.record = record;
-        this.source = record.getStruct(Envelope.FieldName.SOURCE);
-        this.transaction = record.schema().field(Envelope.FieldName.TRANSACTION) != null ? record.getStruct(Envelope.FieldName.TRANSACTION) : null;
-        this.op = record.getString(Envelope.FieldName.OPERATION);
-        this.opSchema = schema.field(Envelope.FieldName.OPERATION).schema();
-        this.ts_ms = record.getInt64(Envelope.FieldName.TIMESTAMP).toString();
-        this.ts_msSchema = schema.field(Envelope.FieldName.TIMESTAMP).schema();
-        this.connectorType = source.getString(AbstractSourceInfo.DEBEZIUM_CONNECTOR_KEY);
-        this.dataSchema = getDataSchema(schema, connectorType, dataFields);
+    protected RecordParser(RecordAndMetadata recordAndMetadata, String... dataFields) {
+        this.record = recordAndMetadata.record();
+        this.source = recordAndMetadata.source();
+        this.transaction = recordAndMetadata.transaction();
+        this.op = recordAndMetadata.operation();
+        this.opSchema = Schema.STRING_SCHEMA;
+
+        if (recordAndMetadata instanceof RecordAndMetadataBaseImpl) {
```

Comment:
Do we need `instanceof` here? Is not something that can be delegated to `RecordAndMetadata` implementation?